File: Medicine_business/medicine-common/medicine-common-redis/src/main/java/com/zhangyichuang/medicine/common/redis/core/RedisZSetCache.java
package com.zhangyichuang.medicine.common.redis.core;

import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.Cursor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.ScanOptions;
import org.springframework.data.redis.core.ZSetOperations;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import java.util.*;
import java.util.concurrent.TimeUnit;

/**
 * Redis 有序集合（ZSet）操作缓存工具类
 * 提供有序集合数据结构的各种操作方法，包括批量扫描功能
 *
 * @author Chuang
 * <p>
 * created on 2025/7/25
 */
@Component
@RequiredArgsConstructor
@SuppressWarnings(value = {"unchecked", "rawtypes"})
public final class RedisZSetCache {

    /**
     * 日志记录器。
     */
    private static final Logger log = LoggerFactory.getLogger(RedisZSetCache.class);

    private final static int scanCount = 1000;
    public final RedisTemplate redisTemplate;

    /**
     * 添加有序集合成员
     *
     * @param key    Redis 键
     * @param member 成员
     * @param score  分值
     * @param <T>    成员类型
     */
    public <T> void zAdd(final String key, final T member, final double score) {
        redisTemplate.opsForZSet().add(key, member, score);
    }

    /**
     * 添加有序集合成员（设置超时时间）
     *
     * @param key     Redis 键
     * @param member  成员
     * @param score   分值
     * @param <T>     成员类型
     * @param timeout 超时时间 单位秒
     */
    public <T> void zAdd(final String key, final T member, final double score, final long timeout) {
        redisTemplate.opsForZSet().add(key, member, score);
        redisTemplate.expire(key, timeout, TimeUnit.SECONDS);
    }

    /**
     * 添加有序集合成员（设置超时时间和单位）
     *
     * @param key      Redis 键
     * @param member   成员
     * @param score    分值
     * @param <T>      成员类型
     * @param timeout  超时时间
     * @param timeUnit 时间单位
     */
    public <T> void zAdd(final String key, final T member, final double score, final long timeout, final TimeUnit timeUnit) {
        redisTemplate.opsForZSet().add(key, member, score);
        redisTemplate.expire(key, timeout, timeUnit);
    }

    /**
     * 设置有序集合有效期(单位秒)
     *
     * @param key     Redis 键
     * @param timeout 有效期
     */
    public void expire(final String key, final long timeout) {
        redisTemplate.expire(key, timeout, TimeUnit.SECONDS);
    }

    /**
     * 获取有序集合有效期
     *
     * @param key Redis 键
     */
    public void expire(final String key, final long timeout, final TimeUnit unit) {
        redisTemplate.expire(key, timeout, unit);
    }


    /**
     * 批量删除有序集合中的一个或多个成员
     *
     * @param key     Redis 键
     * @param members 要删除的成员数组
     * @param <T>     成员类型
     * @return 删除的成员个数
     */
    public <T> Long zRemove(final String key, final T... members) {
        return redisTemplate.opsForZSet().remove(key, members);
    }

    /**
     * 按排名（索引）范围获取有序集合成员
     *
     * @param key   Redis 键
     * @param start 开始索引（从 0 开始）
     * @param end   结束索引（-1 表示最后一个）
     * @param <T>   成员类型
     * @return 成员集合（按分值从小到大排序）
     */
    public <T> Set<T> zRange(final String key, final long start, final long end) {
        return redisTemplate.opsForZSet().range(key, start, end);
    }

    /**
     * 按分值范围获取有序集合成员
     *
     * @param key      Redis 键
     * @param minScore 最小分值
     * @param maxScore 最大分值
     * @param <T>      成员类型
     * @return 成员集合（分值在 [minScore, maxScore] 区间内）
     */
    public <T> Set<T> zRangeByScore(final String key, final double minScore, final double maxScore) {
        return redisTemplate.opsForZSet().rangeByScore(key, minScore, maxScore);
    }

    /**
     * 统计有序集合中分值在给定区间的成员数量
     *
     * @param key      Redis 键
     * @param minScore 最小分值
     * @param maxScore 最大分值
     * @return 成员数量
     */
    public Long zCount(final String key, final double minScore, final double maxScore) {
        return redisTemplate.opsForZSet().count(key, minScore, maxScore);
    }

    /**
     * 获取有序集合的总成员数
     *
     * @param key Redis 键
     * @return 成员数量
     */
    public Long zCard(final String key) {
        return redisTemplate.opsForZSet().zCard(key);
    }


    /**
     * 获取有序集合的所有成员（带分值）
     *
     * @param key Redis 键
     * @return 值集合（带分值）
     */
    public Set<ZSetOperations.TypedTuple<String>> getAllWithScore(String key) {
        return redisTemplate.opsForZSet().rangeWithScores(key, 0, -1);
    }


    /**
     * 获取某个成员的分值
     *
     * @param key    Redis 键
     * @param member 成员
     * @param <T>    成员类型
     * @return 分值；null 表示成员不存在
     */
    public <T> Double zScore(final String key, final T member) {
        return redisTemplate.opsForZSet().score(key, member);
    }

    /**
     * 为有序集合中的某个成员分值加上给定增量
     *
     * @param key    Redis 键
     * @param member 成员
     * @param delta  增量（可以为负数，用于减分）
     * @param <T>    成员类型
     * @return 操作后的新分值
     */
    public <T> Double zIncrementScore(final String key, final T member, final double delta) {
        return redisTemplate.opsForZSet().incrementScore(key, member, delta);
    }


    /**
     * 批量扫描ZSet键 - 使用Redis SCAN操作高效获取匹配的有序集合键
     * 推荐在生产环境中使用此方法，避免阻塞Redis服务器
     *
     * @param keyPattern Redis键模式，支持通配符（如：ranking:*）
     * @return 匹配的键列表，如果没有匹配的键则返回空列表
     */
    public List<String> scanKeys(final String keyPattern) {
        if (!StringUtils.hasText(keyPattern)) {
            return new ArrayList<>();
        }

        List<String> keys = new ArrayList<>();
        try {
            ScanOptions options = ScanOptions.scanOptions()
                    .match(keyPattern)
                    // 每次扫描的数量，可根据实际情况调整
                    .count(scanCount)
                    .build();

            try (Cursor<String> cursor = redisTemplate.scan(options)) {
                while (cursor.hasNext()) {
                    keys.add(cursor.next());
                }
            }

            log.debug("Redis ZSet scan completed, pattern: {}, found {} keys", keyPattern, keys.size());
        } catch (Exception e) {
            log.error("Redis ZSet scan keys failed, pattern: {}, error: {}", keyPattern, e.getMessage(), e);
            // 发生异常时返回空列表，避免影响业务流程
            return new ArrayList<>();
        }

        return keys;
    }

    /**
     * 批量扫描ZSet键和成员 - 获取匹配模式的有序集合键及其所有成员（带分值）
     * 使用Redis SCAN操作高效扫描键，然后批量获取每个ZSet的所有成员和分值
     *
     * @param keyPattern Redis键模式，支持通配符（如：ranking:*）
     * @return Map，键为ZSet键名，值为该ZSet的所有成员（带分值），如果没有匹配的键则返回空Map
     */
    public Map<String, Set<ZSetOperations.TypedTuple<Object>>> scanKeysWithValues(final String keyPattern) {
        if (!StringUtils.hasText(keyPattern)) {
            return new HashMap<>();
        }

        // 首先扫描获取所有匹配的键
        List<String> keys = scanKeys(keyPattern);
        if (keys.isEmpty()) {
            return new HashMap<>();
        }

        Map<String, Set<ZSetOperations.TypedTuple<Object>>> result = new HashMap<>();
        try {
            // 逐个获取每个ZSet的所有成员（带分值）
            for (String key : keys) {
                Set<ZSetOperations.TypedTuple<Object>> zsetMembers =
                        redisTemplate.opsForZSet().rangeWithScores(key, 0, -1);
                result.put(key, zsetMembers);
            }
        } catch (Exception e) {
            log.error("Redis ZSet scan keys with values failed, pattern: {}, error: {}", keyPattern, e.getMessage(), e);
            return new HashMap<>();
        }

        return result;
    }


    /**
     * 使用ZSCAN扫描单个有序集合的成员（带分值）
     * 注意：这是惰性、增量式扫描；本方法会将结果聚合到内存中返回。
     * 如需极致大集合的流式处理，建议提供回调式消费。
     *
     * @param key ZSet键名
     * @return 成员（带分值）集合
     */
    public Set<ZSetOperations.TypedTuple<Object>> zScanWithScores(final String key) {
        if (!StringUtils.hasText(key)) {
            return Collections.emptySet();
        }

        Set<ZSetOperations.TypedTuple<Object>> result = new LinkedHashSet<>();
        try {
            ScanOptions options = ScanOptions.scanOptions()
                    .count(scanCount)
                    .build();
            try (Cursor<ZSetOperations.TypedTuple<Object>> cursor = redisTemplate.opsForZSet().scan(key, options)) {
                while (cursor.hasNext()) {
                    result.add(cursor.next());
                }
            }
        } catch (Exception e) {
            log.error("Redis ZSet scan members failed, key: {}, error: {}", key, e.getMessage(), e);
            return Collections.emptySet();
        }
        return result;
    }


}

File: medicine-ai-agent/app/core/database/redis/redis_zset_cache.py
from __future__ import annotations

from typing import Any

from redis import Redis
from redis.exceptions import RedisError

from app.core.database.redis.config import get_redis_connection
from app.core.database.redis.redis_cache import (
    DEFAULT_SCAN_COUNT,
    _decode_key,
    _raise_redis_operation_error,
)


class RedisZSetCache:
    """Redis ZSet 结构操作封装。"""

    def __init__(self, redis_client: Redis | None = None) -> None:
        """初始化 Redis ZSet 操作封装。

        Args:
            redis_client: 可选 Redis 客户端，未传时使用全局连接。
        """
        self._redis = redis_client or get_redis_connection()

    def z_add(self, key: str, member: Any, score: float) -> int:
        """添加有序集合成员。

        Args:
            key: Redis 键。
            member: 成员值。
            score: 成员分值。

        Returns:
            int: 实际新增成员数量（0 或 1）。

        Raises:
            ServiceException: Redis 写入失败时抛出。
        """
        try:
            return int(self._redis.zadd(key, {member: score}))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zadd", exc=exc) from exc

    def z_add_with_ttl(self, key: str, member: Any, score: float, timeout: int) -> int:
        """添加有序集合成员并设置过期时间（秒）。

        Args:
            key: Redis 键。
            member: 成员值。
            score: 成员分值。
            timeout: 过期秒数。

        Returns:
            int: 实际新增成员数量（0 或 1）。

        Raises:
            ServiceException: Redis 写入或设置过期失败时抛出。
        """
        added = self.z_add(key, member, score)
        self.expire(key, timeout)
        return added

    def z_remove(self, key: str, *members: Any) -> int:
        """删除有序集合成员。

        Args:
            key: Redis 键。
            *members: 待删除成员列表。

        Returns:
            int: 实际删除成员数量。

        Raises:
            ServiceException: Redis 删除失败时抛出。
        """
        if not members:
            return 0
        try:
            return int(self._redis.zrem(key, *members))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zrem", exc=exc) from exc

    def z_range(self, key: str, start: int, end: int) -> list[Any]:
        """按排名范围读取成员。

        Args:
            key: Redis 键。
            start: 起始排名（包含）。
            end: 结束排名（包含），`-1` 表示末尾。

        Returns:
            list[Any]: 成员列表。

        Raises:
            ServiceException: Redis 读取失败时抛出。
        """
        try:
            return list(self._redis.zrange(key, start, end))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zrange", exc=exc) from exc

    def z_range_by_score(self, key: str, min_score: float, max_score: float) -> list[Any]:
        """按分值范围读取成员。

        Args:
            key: Redis 键。
            min_score: 最小分值（包含）。
            max_score: 最大分值（包含）。

        Returns:
            list[Any]: 分值区间内成员列表。

        Raises:
            ServiceException: Redis 读取失败时抛出。
        """
        try:
            return list(self._redis.zrangebyscore(key, min_score, max_score))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zrangebyscore", exc=exc) from exc

    def z_count(self, key: str, min_score: float, max_score: float) -> int:
        """统计分值区间内成员数。

        Args:
            key: Redis 键。
            min_score: 最小分值（包含）。
            max_score: 最大分值（包含）。

        Returns:
            int: 区间成员数量。

        Raises:
            ServiceException: Redis 统计失败时抛出。
        """
        try:
            return int(self._redis.zcount(key, min_score, max_score))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zcount", exc=exc) from exc

    def z_card(self, key: str) -> int:
        """读取集合总成员数。

        Args:
            key: Redis 键。

        Returns:
            int: 成员总数。

        Raises:
            ServiceException: Redis 查询失败时抛出。
        """
        try:
            return int(self._redis.zcard(key))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zcard", exc=exc) from exc

    def get_all_with_score(self, key: str) -> list[tuple[Any, float]]:
        """读取所有成员及分值。

        Args:
            key: Redis 键。

        Returns:
            list[tuple[Any, float]]: 由成员与分值组成的列表。

        Raises:
            ServiceException: Redis 读取失败时抛出。
        """
        try:
            return list(self._redis.zrange(key, 0, -1, withscores=True))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zrange_withscores", exc=exc) from exc

    def z_score(self, key: str, member: Any) -> float | None:
        """读取成员分值。

        Args:
            key: Redis 键。
            member: 成员值。

        Returns:
            float | None: 成员分值，不存在返回 `None`。

        Raises:
            ServiceException: Redis 查询失败时抛出。
        """
        try:
            score = self._redis.zscore(key, member)
            return float(score) if score is not None else None
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zscore", exc=exc) from exc

    def z_increment_score(self, key: str, member: Any, delta: float) -> float:
        """对成员分值做增量。

        Args:
            key: Redis 键。
            member: 成员值。
            delta: 分值增量。

        Returns:
            float: 增量后的分值。

        Raises:
            ServiceException: Redis 自增失败时抛出。
        """
        try:
            return float(self._redis.zincrby(key, delta, member))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zincrby", exc=exc) from exc

    def expire(self, key: str, timeout: int) -> bool:
        """设置键过期时间（秒）。

        Args:
            key: Redis 键。
            timeout: 过期秒数。

        Returns:
            bool: 设置成功返回 `True`，否则返回 `False`。

        Raises:
            ServiceException: Redis 设置过期失败时抛出。
        """
        try:
            return bool(self._redis.expire(key, timeout))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="expire", exc=exc) from exc

    def scan_keys(self, key_pattern: str, *, count: int = DEFAULT_SCAN_COUNT) -> list[str]:
        """按模式扫描 ZSet 键列表。

        Args:
            key_pattern: 键匹配模式。
            count: 每次 SCAN 的建议批次大小。

        Returns:
            list[str]: 匹配到的键列表。

        Raises:
            ServiceException: Redis 扫描失败时抛出。
        """
        cursor = 0
        keys: list[str] = []
        try:
            while True:
                cursor, batch = self._redis.scan(
                    cursor=cursor,
                    match=key_pattern,
                    count=count,
                )
                keys.extend(_decode_key(item) for item in batch)
                if int(cursor) == 0:
                    break
            return keys
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="scan", exc=exc) from exc

    def scan_keys_with_values(
            self,
            key_pattern: str,
            *,
            count: int = DEFAULT_SCAN_COUNT,
    ) -> dict[str, list[tuple[Any, float]]]:
        """按模式扫描 ZSet 键并读取成员及分值。

        Args:
            key_pattern: 键匹配模式。
            count: 每次 SCAN 的建议批次大小。

        Returns:
            dict[str, list[tuple[Any, float]]]: 键与成员分值列表映射。

        Raises:
            ServiceException: Redis 扫描或读取失败时抛出。
        """
        matched_keys = self.scan_keys(key_pattern, count=count)
        if not matched_keys:
            return {}
        result: dict[str, list[tuple[Any, float]]] = {}
        try:
            for key in matched_keys:
                result[key] = list(self._redis.zrange(key, 0, -1, withscores=True))
            return result
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="scan_zset_values", exc=exc) from exc

    def z_scan_with_scores(self, key: str, *, count: int = DEFAULT_SCAN_COUNT) -> list[tuple[Any, float]]:
        """扫描单个 ZSet 成员并携带分值。

        Args:
            key: Redis 键。
            count: 每次 ZSCAN 的建议批次大小。

        Returns:
            list[tuple[Any, float]]: 成员与分值列表。

        Raises:
            ServiceException: Redis 扫描失败时抛出。
        """
        cursor = 0
        result: list[tuple[Any, float]] = []
        try:
            while True:
                cursor, batch = self._redis.zscan(
                    key,
                    cursor=cursor,
                    count=count,
                )
                result.extend((member, float(score)) for member, score in batch)
                if int(cursor) == 0:
                    break
            return result
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="zscan", exc=exc) from exc
